packages/interactive-messages/src/adapter.ts

```diff
@@ -0,0 +1,637 @@
+    throw new TypeError('Constraints cannot be undefined or null');
+  }
+  if (!isPlainObject(matchingConstraints)) {
+    ret.callbackId = matchingConstraints as string | RegExp;
+  } else {
+    ret = Object.assign({}, matchingConstraints as BaseConstraints);
+  }
+  return ret;
+}
+
+/**
+ * Validates general properties of a matching constraints object
+ * @param matchingConstraints - object describing the constraints on a callback
+ * @returns `false` represents successful validation, an error represents failure and describes why validation failed.
+ */
+function validateConstraints(matchingConstraints: BaseConstraints): Error | false {
+  if (!isFalsy(matchingConstraints.callbackId) &&
+      !(isString(matchingConstraints.callbackId) || isRegExp(matchingConstraints.callbackId))) {
+    return new TypeError('Callback ID must be a string or RegExp');
+  }
+
+  if (!isFalsy(matchingConstraints.blockId) &&
+    !(isString(matchingConstraints.blockId) || isRegExp(matchingConstraints.blockId))) {
+    return new TypeError('Block ID must be a string or RegExp');
+  }
+
+  if (!isFalsy(matchingConstraints.actionId) &&
+    !(isString(matchingConstraints.actionId) || isRegExp(matchingConstraints.actionId))) {
+    return new TypeError('Action ID must be a string or RegExp');
+  }
+
+  return false;
+}
+
+/**
+ * Validates properties of a matching constraints object specific to registering an options request
+ * @param matchingConstraints - object describing the constraints on an options handler
+ * @returns `false` represents successful validation, an error represents failure and describes why validation failed.
+ */
+function validateOptionsConstraints(optionsConstraints: OptionsConstraints): Error | false {
+  if (!isFalsy(optionsConstraints.within) &&
+    !(optionsConstraints.within === 'interactive_message' ||
+      optionsConstraints.within === 'block_actions' ||
+      optionsConstraints.within === 'dialog')
+  ) {
+    return new TypeError('Within must be \'block_actions\', \'interactive_message\' or \'dialog\'');
+  }
+
+  // We don't need to validate unfurl, we'll just cooerce it to a boolean
+  return false;
+}
+
+/**
+ * An adapter for Slack's interactive message components such as buttons, menus, and dialogs.
+ * @typicalname slackInteractions
+ */
+export class SlackMessageAdapter {
+  /**
+   * Slack app signing secret used to authenticate request
+   */
+  public signingSecret: string;
+
+  /**
+   * The number of milliseconds to wait before flushing a syncrhonous response to an incoming request and falling back
+   * to an asynchronous response.
+   */
+  public syncResponseTimeout: number;
+
+  /**
+   * Whether or not promises that resolve after the syncResponseTimeout can fallback to a request for the response_url.
+   * This only works in cases where the semantic meaning of the response and the response_url are the same.
+   */
+  public lateResponseFallbackEnabled: boolean;
+
+  private callbacks: ConstrainedHandler[];
```

Comment:
(the following comment isn't just about this line, but its as good a place to anchor it as any I could find)

I see a few code smells in the way the types have been modeled.

1. The tuple type `ConstrainedHandler` is never used (except where its used to describe the type of the `callbacks` instance variable) without indexing for a certain position (this holds true for each of the constituent types in the union). It seems like its individual items are more important to the APIs and implementation than the tuple type.
2. The return type of `formatMatchingConstraints()` doesn't seem to matter because its always erased (casted away).

Here is a counter-proposal (in a playground so we can see how the type checker evaluates it):

https://www.typescriptlang.org/play/index.html#code/MYGwhgzhAEDK7ANYFkCmUwHNUEEAmYADgC6oBO0A3gFDTQD090AdK7dIWQJYBuYp0YGBAgARmCQQAXNADasYgHsyqPAGFFAOwjEyYLpuIQAPDk0BPDdt37DEAHwAaaGuFiJiALqzPAbnaMLGx0wFo6ZACuwEpkABQAlFTsdHTEABZcEMxCIuKS0AC8cn7sAL4BTKzM7OyEEaIgXMDQEsRcWrHJ0AC2-MAZmphW4bZGMuEGmNAAPtAASqiYAKIAHoQz0DjR7ZrDNgZGjl057kgyW21aABJgmngg5Ed08TLpmUkpKYEAMoqKiD0wM5NIpBJA2oMAIRdEJhYjQABmyl6xFI6jhegOMCKSLIKOQfQGQwxowgsRR-Ume0xdni-k+sOs0B0yjADVQ1NJhWgAHlRAArVDRZiQCBcTCaWK4lFozlY5yUaBpW73cgAFXMhFQMgA5K0djqWjBQkzSnSYQxKmwLYEAFIRHTQJTMtKKADuTuV8PSqCNYolbK4jWI5kEaSFiBgigi8NiBk9vpUwkEijwvreMG6qHSqbBIhgGeYKkwmVIZFcuQ8CXiFsLJzykeYdQgaVishZenZcrsznrHk85oZ0BUxAiZE0nsy9Og5Vq9UazUUJB2ZK6FKJ3bGzN0kw2C2Waw2POXYU3ECeKT7Z0222uKoeZAvL0nMBoQ5Njul-FlJKx3K-xAEsQlKDGe5KElSv60tOnwfvCHZsg8Z7cnygrCqK4qSgBP7WDShxUEq97qpq2rQDqS6XNohqQCmpqDgygRVLWGRZFejbNq27YxIhHJQfhbEDjBKQjmOE4ZtOs50IECzdIoPDoMyihZtAigIgmzJZiIiJkEp0AGGmKy7gcoI+k6ESEA8TokbU3B8AIxaluQFanIgsRwXh0hwDEqhnqYFhnk4eYuTIzkNs+GYfAxVrVEOInji+EnUBU0BqixmwAAoAJLQNgRjMo0mBpMQIChhEmCNI80CiDGemTpoAJcHlbrKOk6k2GmqkIjZvD8L666hVWhBgOYICKGAeAyLc5jPvI3norhpJ+ZYfEOM4A1IJ4GwRHcqAIgYqiRZ8NpMJlYlpVmOZ4L2Wh4I1K7MlEaRGtAAAG7mkswyp3A+GpaoUBRFORJ5US9LQqNADqqDFUXQAA6r6QgTioFkSOm4bdNAbqNU9wAOkoGMhn9mARGAZB4DAEAmV6r3vVioOI4R8ltSRvLAxAyHxlovqiHomj9CpammSat2UUlQ6McEDJxROsiKl9qpkL9pF6remiGqUzgJIU9hUKUJR0OUs6BBcOycnYLR3KzlHsytYO+kNZDwqp6l1A0TQZZlzBwwj0YgHgYZ-BA6agqgKw2NE6lit0Qak9QgScEu5BtAp6T8AmoYtqTvrxqE3SyROxkW3phjkAiqOYxk-M+qGDOiIorWh6QdzONV3rhioOowCCLR4HwhhYNnE54Iou6U9UgRK7AwDcCQKbdA7KckTA8ZgNukTRGOyaxHoPoUKnE4gtHmjCIkbrDYpCYd13oIYQGohBo1oa3QiCLkKgfPoF7mVqW6vq-+DzpG56AjrcYuZYy7AFQEcQIqd4QHFLuXZqEQ-bQE0KgeSFBRC+khv7LQek8rugnLiC2oYcBZWgAAWnwZjX2-t+R42HKgWSTNKbKWRm0SiVUappyFooR0zt9JcBUKrREygWgQyDhQN0rpCIQDjkwZ0o1-gQ3WK6D0qBsbkAFgmIOYCEGQJgFnaAaY9poLwNUagodCAtT0WQCBvoTZaGQm+S8bgGyZTwAAfnGDuQYe5FirEIEJBoigkAeO8WvXcsx9yBKEvqLQ4SfHcD8dEgJax-BdEJqgCJExBj+FnFYmx8C7Hl2PNbZxxw3EeESZElJ8w0lBK6CEsJXiklRPqQeRpnx4maBqbkqYqTOkZM+FjN4mgIk6maYgAA+j0iAhpZg6mKfqeS0yswYGwAssit1hCKEwDqags4smbH8rbIojjdi21mGUlcZ4MmBAUMoHyts0wQGnlwLBBZwy2OPiAbcrJsCiLxGnZ2U1aIjAOFZLU1BjmPJUPNCFdhjBqGgI3d+5MTnLQWliHWRQUUADJDqEW+sRLUuoelbKBtbHUQlAgeNQG4cwzhf4pmQf7SG4jhZ3S0MmY5zoICjKrt8454gg64LOoPOo8JabmyyReKSTBW5VSFGADloLoAIABM7UOrQoXoGcKZdBwgSacOHugNW8JmpkABKZM+oYWVn0MNQ50VrEDQwVcShWSsQqorDuimANzTy208WRCiK5DRSC6IEIcMaUVoruDAC5yEQ0q0ohGqNTAY1DjQRg-JSUETbREQBICIFiTYqRcheNGKzBYsRUYewnRPjrkguWrc-T-GdI2Geagz4KmfC4GpcCwENxnIBhDHapiDrTFmM20Co6iiaGQSARILihzpB0h6NBHolZLDIDpOIOpkKIxBPCLB46TH7VwRQRdIgdT0QNl0B48IRznFOa27EBFShCQHdAWIWTnazrLXW7EY7O6+P2TMGdEE53vuLjoW4kDnYxLWCui0dARzZCqS07kgG7ldFKKikAujV1S2zChAUQpiAiigJhWIlANaAmHS24D96ZxdGlow+ENE8NGyYKldAvojGhEIFwA6cktEYPEG0boMKWYLAgNYy2RRyToAgAPSaFhEgFB1ulHS0cg7GCmvYfwxyLk3BJRQZTQ0RpjQmiQ5wKgFM3RkPJxTeAtM6zBbMXTSlMioEMxYHWsxtoXrMSZlmgbtDmYVtyWI1nRrjQ09NbWJCNg+f0-5ozW0J2XvC39daAJzmq2iw+I8bMSvkAyUAA

Reply:
Thank you for writing this all this out! 🤩 I agree that there was definitely spaghetti types going on before. I think half of it was me not wanting to change *any* of the functionality of the code.

All of your occurrences of `StoredConstraints` was used with `AnyConstraints` as the generic type, so I merged the two into a union type. This made the expanded signature of `matchCallback` to be more graceful. Because of it being a union instead I didn't really have to change the body of this function anymore to guard the typechecker into figuring out what type of constraints were being used.

As a side effect, this also drops the `lodash.isplainobject` dependency. 🎉

Reply:
Sounds good! Would love to see the result of this update.